types/hapi__joi/index.d.ts

```diff
@@ -152,1245 +226,1922 @@ export interface DomainOptions {
+         * For example, `Joi.number().min(1).rule({ keep: true }).min(2)` will keep both `min()` rules instead of the later rule overriding the first.
+         *
+         * @default false
+         */
+        keep?: boolean;
 
-export interface FunctionSchema extends AnySchema {
-    /**
-     * Specifies the arity of the function where:
-     * @param n - the arity expected.
-     */
-    arity(n: number): this;
-
-    /**
-     * Specifies the minimal arity of the function where:
-     * @param n - the minimal arity expected.
-     */
-    minArity(n: number): this;
-
-    /**
-     * Specifies the minimal arity of the function where:
-     * @param n - the minimal arity expected.
-     */
-    maxArity(n: number): this;
-
-    /**
-     * Requires the function to be a Joi reference.
-     */
-    ref(): this;
-}
+        /**
+         * a single message string or a messages object where each key is an error code and corresponding message string as value.
+         *
+         * The object is the same as the messages used as an option in `any.validate()`.
+         * The strings can be plain messages or a message template.
+         */
+        message?: string | LanguageMessages;
 
-export interface AlternativesSchema extends AnySchema {
-    try(types: SchemaLike[]): this;
-    try(...types: SchemaLike[]): this;
-    when(ref: string | Reference, options: WhenOptions): this;
-    when(ref: Schema, options: WhenSchemaOptions): this;
-}
+        /**
+         * if true, turns any error generated by the ruleset to warnings.
+         */
+        warn?: boolean;
+    }
+
+    interface JoiObject {
+        isJoi: boolean;
+    }
+
+    interface ErrorReport extends Error {
+        code: string;
+        flags: Record<string, ExtensionFlag>;
+        path: string;
+        prefs: ValidationOptions;
+        messages: LanguageMessages;
+        state: State;
+        value: any;
+    }
+
+    interface ValidationError extends Error, JoiObject {
+        name: 'ValidationError';
 
-export interface LazySchema extends AnySchema {
-}
+        /**
+         * array of errors.
+         */
+        details: ValidationErrorItem[];
 
-export interface Reference extends JoiObject {
-    (value: any, validationOptions: ValidationOptions): any;
-    isContext: boolean;
-    key: string;
-    path: string;
-    toString(): string;
-}
+        /**
+         * function that returns a string with an annotated version of the object pointing at the places where errors occurred.
+         * @param stripColors - if truthy, will strip the colors out of the output.
+         */
+        annotate(stripColors?: boolean): string;
+
+        _object: any;
+    }
+
+    interface ValidationErrorItem {
+        message: string;
+        path: Array<string | number>;
+        type: string;
+        context?: Context;
+    }
+
+    type ValidationErrorFunction = (errors: ValidationErrorItem[]) => string | ValidationErrorItem | Error;
+
+    interface ValidationResult {
```

Comment:
I think we should make this generic: 
``
    interface ValidationResult<T = any> {
        error: ValidationError;
        errors: ValidationError;
        warning: ValidationError;
        value: T;
    }
``

Reply:
Mhm, not sure what's best. 
@SimonSchick what do you think?

Reply:
`no-unnecessary-generics` is _usually_ right, as you are effectively substituting casts with generics, which in the end reduces type safety, unless your intent it to make schemas generic and default to `unknown` then this should NOT be generic.

Reply:
But how would it look like then? Am I supposed to cast every validation result like this?
```ts
const result = schema.validate() as {
  error: ValidationError;
  errors: ValidationError;
  warning: ValidationError;
  value: {
    validatedString: string          
  };
}
```

Or is there a simpler way?

Reply:
After reading the comment by @SimonSchick, maybe `unknown` is an OK compromise since it's at least safer than `any`. Supporting a generic still makes sense to me though since it doesn't violate the lint rule since it was used in the params and return value AFAICT.

Reply:
That's the thing, the input value may NOT be the return value since joi performs cohersion/conversion.

Reply:
Exactly, that's why the generic is required you need to specify the return value type. So I *think* we agree here?

Reply:
No, you input `unknown` and you get `unknown`, the generic would do nothing but hide the type assertion, which `no unnecessary generic` attempts to prevent.

Reply:
Agreed you input `unknown`, but returning `unknown` is a choice being made to satisfy the lint rule AFAICT. 

We know what'll be returned is either:

* an object with the `value` key with value `T` (since Joi validated it)
* an object with an `error` key and associated value (if validation failed)

Changing the `any` to `unknown` is probably a good call since changing the lint rule is probably a no-go and it would force developers to apply the type cast?